lib/util/cache/repository/impl/s3.spec.ts

```diff
@@ -0,0 +1,133 @@
+import { Readable } from 'stream';
+import {
+  GetObjectCommand,
+  GetObjectCommandInput,
+  PutObjectCommand,
+  PutObjectCommandInput,
+  PutObjectCommandOutput,
+  S3Client,
+} from '@aws-sdk/client-s3';
+import { mockClient } from 'aws-sdk-client-mock';
+import { partial } from '../../../../../test/util';
+import { GlobalConfig } from '../../../../config/global';
+import { logger } from '../../../../logger';
+import { parseS3Url } from '../../../s3';
+import type { RepoCacheRecord } from '../types';
+import { CacheFactory } from './cache-factory';
+import { RepoCacheS3 } from './s3';
+
+function createGetObjectCommandInput(
+  repository: string,
+  url: string
+): GetObjectCommandInput {
+  return {
+    Bucket: parseS3Url(url)?.Bucket,
+    Key: `github/${repository}/cache.json`,
+  };
+}
+
+function createPutObjectCommandInput(
+  repository: string,
+  url: string,
+  data: RepoCacheRecord
+): PutObjectCommandInput {
+  return {
+    ...createGetObjectCommandInput(repository, url),
+    Body: JSON.stringify(data),
+    ContentType: 'application/json',
+  };
+}
+
+/*
+ * Note: MockedClient.on(Command, input) will match input (using Sinon matchers) to the actual
+ *       'new Command(actualInput)' call within the tested code segment.
+ */
+
+describe('util/cache/repository/impl/s3', () => {
+  const s3Mock = mockClient(S3Client);
+  const repository = 'org/repo';
+  const repoCache = partial<RepoCacheRecord>({ payload: 'payload' });
+  const url = 's3://bucket-name';
+  const err = new Error('error');
+  let getObjectCommandInput: GetObjectCommandInput;
+  let putObjectCommandInput: PutObjectCommandInput;
+  let s3Cache: RepoCacheS3;
+
+  beforeEach(async () => {
+    GlobalConfig.set({ platform: 'github' });
+    jest.clearAllMocks();
+    s3Mock.reset();
+    // every test function should have its own s3 instance for it to work
+    s3Cache = new (await import('./s3')).RepoCacheS3(repository, url);
```

Comment:
i don't think we need a dynamic import in this test file, as we create a new class instance anyways .

also without `jest.resetModules` you'll always get same module from jest cache.

so no benefit for dynamic import here. use static import instead.

Reply:
previously I had issues when using the same s3Client which is a singleton. therefore I had to mock its wrapper's return value using a spy.

this has been fixed since and is not an issue anymore, i.e. the same singleton s3 client can be used.